doc/sonic-vrf-hld-v0.2.md

```diff
@@ -0,0 +1,343 @@
+
+//enable VRF
+
+$ ip link set dev vrf-blue up
+
+// disable global VRF lookup
+
+$ ip [-6] route add table 10 unreachable default
+
+//binding sw1p3 device to vrf-blue
+
+$ ip link set dev sw1p3 master vrf-blue
+
+// descend local table pref
+
+ip [-6] rule add pref 32765 table local && ip [-6] rule del pref 0
+```
+
+4.  SAI VRF support
+
+SAI right now does not seem having VRF concept, it does have VR.
+
+We propose to implement VR as “virtual router” and VRF as “virtual router
+forwarding”
+
+VR is defined as an logical routing system. VRF is defined as forwarding
+domain within a VR.
+
+As this stage, we assume one VR per system. Only implement VRFs within this
+VR.
+
+Accordingly, we need to add vrf_id to sai_Route_entry and add vrf attribute
+to sai_routeInterface object.
+
+An alternative method is using VR as VRF. But it is needed to add two
+attribution to VR object.
+
+```
+/*
+ * \@brief if it is global vrf
+ *
+ * \@type bool
+ * \@flags CREATE_AND_SET
+ * \@default true
+ */
+ SAI_VIRTUAL_ROUTER_ATTR_GLOBAL
+
+/*
+ * \@brief continue to do global fib lookup while current vrf fib lookup
+ *  missed
+ * 
+ * \@type bool
+ * \@flags CREATE_AND_SET
+ * \@default false
+ */
+ SAI_VIRTUAL_ROUTER_ATTR_FALL_THROUGH
+```
+
+SONiC system diagram for VRF
+============================
+
+The following is high level diagram of modules with VRF support.
+
+## The schema changes
+
+1. Adding VRF related configuration in config_db.json
+
+```
+"VRF": {
+    "VRF-blue": {
+        "fall_through":"1" //enable global fib lookup while vrf fib lookup missed
+    },
+    "VRF-red":{
+        "fall_through": "1"
+    },
+    "VRF-yellow":{
+        "fall_through":"0" //disable global fib lookup while vrf fib lookup missed
+    }
+},
+
+"INTERFACE":{
+    "Ethernet0":{
+        "mtu":"1500",
+        "vrf":"vrf-blue"
+    },
+    "Ethernet1":{
+        "mtu":"1500”,
+        "vrf":"vrf-red"
+    },
+    "Ethernet0|11.11.11.1/24": {},
+    "Ethernet1|12.12.12.1/24": {},
+    "Ethernet2|13.13.13.1/24": {},
+    "Ethernet3|14.14.14.1/24": {},
+},
+
+"VLAN_INTERFACE": {
+    "Vlan100":{
+        "mtu":"1500",
+        "vrf":"vrf-blue"
+    },
+    "Vlan100|15.15.15.1/24": {},
+}
+```
+
+2. **Adding a VRF_TABLE** in APP_DB
+
+```
+;defines virtual routing forward table
+;
+;Status: stable
+
+key = VRF_TABLE:VRF_NAME ;
+fall_through = "1"/"0"
+```
+
+3. **Breaking up app-intf-table into app-intf-table and app-intf-prefix-table**
+
+app-intf-table is defined as the following:
+
+```
+;defines logical network interfaces, an attachment to a PORT name
+;
+;Status: stable
+
+key = INTF_TABLE:ifname
+if_mtu = 1\*4DIGIT ; MTU for the interface
+VRF_NAME = 1\*64VCHAR ;
+```
+
+app-intf-prefix-table is defined as the following:
+
+```
+;defines logical network interfaces with IP-prefix, an attachment to a PORT and
+list of 0 or more ip prefixes;
+
+;Status: stable
+key = INTF_TABLE:ifname:IPprefix ; an instance of this key will be repeated for
```

Comment:
What is the advantage of having multiple keys?

Please provide the schema for this.

Reply:
@nikos-github There is two reason why break up app-intf-table into app-intf-table and app-intf-prefix-table.
1. Multiple ip addresses can be configured on one interface.So I put interface common attribute into app-intf-table and keep ip-prefix specific attribute on app-intf-prefix-table.
2. Interface can be put to specific VRF before ip address is configured on it.

Reply:
jason file looks like below:
    "Ethernet1":{
        "mtu":"1500”,
        "vrf":"vrf-red"
    },
    "Ethernet0|11.11.11.1/24": {},
    "Ethernet1|12.12.12.1/24": {},
    "Ethernet2|13.13.13.1/24": {},
    "Ethernet3|14.14.14.1/24": {},